crates/boundless-market/src/contracts/mod.rs

```diff
@@ -553,20 +567,249 @@ impl Requirements {
+            FulfillmentData::None => None,
+            FulfillmentData::ImageIdAndJournal(_, journal) => Some(journal),
+        }
+    }
+
+    /// Returns the image id if available
+    pub fn image_id(&self) -> Option<Digest> {
+        match self {
+            FulfillmentData::None => None,
+            FulfillmentData::ImageIdAndJournal(image_id, _) => Some(*image_id),
+        }
+    }
+}
+
+impl From<FulfillmentDataImageIdAndJournal> for FulfillmentData {
+    fn from(value: FulfillmentDataImageIdAndJournal) -> Self {
+        Self::ImageIdAndJournal(Digest::from_bytes(*value.imageId), value.journal)
+    }
+}
+
+impl Fulfillment {
+    /// Decode and return the [FulfillmentData] for the this fulfillment.
+    pub fn data(&self) -> Result<FulfillmentData, FulfillmentDataError> {
+        FulfillmentData::decode_with_type(self.fulfillmentDataType, &self.fulfillmentData)
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+/// Errors related to predicate encoding/decoding and evaluation
+pub enum PredicateError {
+    /// Data was not correctly constructed
+    #[error("malformed predicate data")]
+    DataMalformed,
+    /// Invalid predicate type
+    #[error("invalid predicate type")]
+    InvalidType,
+}
+
+#[derive(Clone, Debug)]
+#[non_exhaustive]
+/// A Predicate is a function over the claim that determines whether it meets the clients requirements.
+pub enum Predicate {
+    /// The image id and digest must match the provided image id and journal digest.
+    DigestMatch(Digest, Digest),
+    /// The image id and journal prefix must match the provided image id and journal.
+    PrefixMatch(Digest, Bytes),
+    /// The claim digest must match the provided claim digest.
+    ClaimDigestMatch(Digest),
+}
+
+impl From<Predicate> for RequestPredicate {
+    fn from(value: Predicate) -> Self {
+        match value {
+            Predicate::DigestMatch(image_id, journal_digest) => Self {
+                predicateType: PredicateType::DigestMatch,
+                data: [image_id.as_bytes(), journal_digest.as_bytes()].concat().into(),
+            },
+            Predicate::PrefixMatch(image_id, journal_prefix) => Self {
+                predicateType: PredicateType::PrefixMatch,
+                data: [image_id.as_bytes(), journal_prefix.as_ref()].concat().into(),
+            },
+            Predicate::ClaimDigestMatch(claim_digest) => Self {
+                predicateType: PredicateType::ClaimDigestMatch,
+                data: claim_digest.as_bytes().to_vec().into(),
+            },
+        }
+    }
+}
+
+impl TryFrom<RequestPredicate> for Predicate {
+    type Error = PredicateError;
+
+    fn try_from(value: RequestPredicate) -> Result<Self, Self::Error> {
+        match value.predicateType {
+            PredicateType::DigestMatch => {
+                // Don't need to check the length because that is checked when converting to Digest
+                let (image_id, journal_digest) = value.data.as_ref().split_at(32);
```

Comment:
This will panic if the predicate data is less than 32 bytes. This could actually be meaningful in that an attacker can submit malformed data to the network that will cause the Broker to have a panic in the task. A the very least, this would be a performance issue. Let's use `.split_at_checked` here and below.
```suggestion
                let (image_id, journal_digest) = value.data.as_ref().split_at_checked(32).ok_or(PredicateError::DataMalformed)?;
```